File: dart/inventory/inventory.dart
class InventoryItem {
  String id;
  String name;
  int price;

  InventoryItem(this.id, this.name, this.price);
}

enum StoreEventType {
  PURCHASED,
  REFUNDED
}

class StoreEvent {
  StoreEventType eventType;
  int timestamp = DateTime.now().millisecondsSinceEpoch;
  var payload;

  StoreEvent(this.eventType, this.payload);
}

class Store {
  String name;
  int openingTime = 900;
  int closingTime = 2100;
  Map<String, int> stock = {};
  List<StoreEvent> events = [];

  Store(this.name);
  
  updateItemQuantity(InventoryItem item, int amount) {
    stock.update(item.id, (value) => amount, ifAbsent: () => amount);
  }

  purchaseItem(InventoryItem item) {
    var currentItemStock = stock[item.id] ?? 0;
    if (currentItemStock <= 0) return;
    events.add(StoreEvent(StoreEventType.PURCHASED, item.id));
    updateItemQuantity(item, currentItemStock - 1);
  }

  refundItem(InventoryItem item) {
    var currentItemStock = stock[item.id] ?? 0;
    events.add(StoreEvent(StoreEventType.REFUNDED, item.id));
    updateItemQuantity(item, currentItemStock + 1);
  }
}

main() {
  var item1 = InventoryItem("1", "Item 1 - A", 500);
  var item2 = InventoryItem("2", "Item 2 - B", 750);
  var item3 = InventoryItem("3", "Item 3 - C", 1000);
  
  var store = Store("Bob's Shop");

  store.updateItemQuantity(item1, 10);
  store.updateItemQuantity(item2, 7);
  store.updateItemQuantity(item3, 5);

  store.purchaseItem(item1);

  print(store.stock);
}

File: cpp/snippets/inventory.cpp
#include <iostream>
#include <string>
#include <map>

struct InventoryItem {
  std::string id;
  std::string name;
  unsigned int price;
  InventoryItem(std::string _id, std::string _name, unsigned int _price) : id(_id), name(_name), price(_price) {}
};

enum StoreEventType
{
  PURCHASED,
  REFUNDED
};

struct StoreEvent
{
  StoreEventType eventType;
  unsigned int timestamp;
  std::string payload;
  StoreEvent(StoreEventType _eventType, std::string _payload) : eventType(_eventType), payload(_payload) {
    timestamp = 0;
  }
};

class Store
{
  std::string name;
  unsigned int openingTime;
  unsigned int closingTime;
  std::map<std::string, unsigned int> stock;
  std::vector<StoreEvent> events;

public:
  Store(std::string _name) : name(_name) {
    openingTime = 900;
    closingTime = 1700;
  }
  
  void update_item_quantity(InventoryItem& item, unsigned int amount) {
    stock[item.id] = amount;
  }
  
  void purchase_item(InventoryItem& item) {
    if (stock[item.id] <= 0) {
      return;
    }

    events.push_back(StoreEvent(PURCHASED, item.id));
    update_item_quantity(item, stock[item.id] - 1);
  }
  
  void refund_item(InventoryItem& item) {
    events.push_back(StoreEvent(REFUNDED, item.id));
    update_item_quantity(item, stock[item.id] + 1);
  }
  
  void get_store_status() {
    for (const auto& entry : stock) {
      std::cout << entry.first << ": " << entry.second << std::endl;
    }
  }
};

int main() {
  InventoryItem item1 = InventoryItem("1", "Item 1 - A", 500);
  InventoryItem item2 = InventoryItem("2", "Item 2 - B", 750);
  InventoryItem item3 = InventoryItem("3", "Item 3 - C", 1000);

  Store store = Store("Bob's Shop");
  
  store.update_item_quantity(item1, 10);
  store.update_item_quantity(item2, 7);
  store.update_item_quantity(item3, 5);

  store.purchase_item(item1);

  store.get_store_status();
}

// Local Variables:
// compile-command: "g++ -std=c++11 -Wall -Wextra -fsanitize=address -o inventory inventory.cpp && ./inventory"
// End:
